---
title: "scRNA Analysis Using Seurat"
output: html_notebook
---

```{r}
# load necessary packages
library(dplyr) # 5.3.0
library(Seurat) # 1.1.4
library(patchwork) # 1.3.2
```

```{r}
# Set your working directory
setwd("C:\\scRNA_analysis_using_Seurat")
getwd()
```
# Setup seurat objects

```{r}
# 1. load count data (cellranger ouput,  h5 file format)
donor1_counts <- Read10X_h5("data/Donor1_filtered_feature_bc_matrix.h5")
donor2_counts <- Read10X_h5("data/Donor2_filtered_feature_bc_matrix.h5")
donor3_counts <- Read10X_h5("data/Donor3_filtered_feature_bc_matrix.h5")

```

```{r}
donor1 <- CreateSeuratObject(
  counts = donor1_counts,
  min.cells = 3,   # keep genes expressed in >= 3 cells
  min.features = 200   # keeps only cells that have at least 200 unique genes detected.
)
donor1
```

```{r}

Cells(donor1) # get cell barcodes
Features(donor1) # get feature names
donor1[["RNA"]]$counts # Get counts/expression matrix
```


```{r}
donor2 <- CreateSeuratObject(
  counts = donor2_counts,
  min.cells = 3,   
  min.features = 200   
)

donor3 <- CreateSeuratObject(
  counts = donor3_counts,
  min.cells = 3,   
  min.features = 200   
)

```

```{r}
donor1$sample <- "donor1"
head(donor1[[]]) # get meta data
```

```{r}
donor2$sample <- "donor2"
donor3$sample <- "donor3"

```

```{r}
# Combine seurat objects
combined <- merge(donor1, y = list(donor2, donor3), add.cell.ids = c("donor1", "donor2", "donor3"))

dim(combined)

```
# STANDARD PRE-PROCESSING WORKFLOW

```{r}
# Quality Control and selecting cells for further analysis
# QC metrics commonly used include:
# 1. Number of unique genes detected = count of genes with ≥1 read/UMI 
# 2. Total reads (or total UMIs) = all reads/UMIs in the cell 
# 3. Percentage of reads that map to the mitochondrial genome

# meta.data stores cell-level metadata (one row per cell)
# nFeature_RNA stores unique genes/cells
# nCount_RNA stores total reads/cells
```

```{r}
# Calculates the percentage of reads/UMIs from mitochondrial genes for each cell.
# Stores the result as a numeric column (percent.mt) in the Seurat object’s metadata.

# In human datasets, mitochondrial genes usually start with MT-.
# For mouse datasets, use mt- instead.

combined[["percent.mt"]] <- PercentageFeatureSet(combined, pattern = "^MT-")
head(combined[[]])

```



```{r}
# Visualize QC metrics as a violin plot
VlnPlot(combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3) 
```

```{r}

plot1 <- FeatureScatter(combined, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(combined, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2 # patchwork combines them
```

```{r}
#Fitler cells
combined <- subset(combined, subset = nFeature_RNA > 200 & nFeature_RNA < 7500 & percent.mt < 20)
dim(combined)

```

```{r}
# Normalizing the data
# log1p( (counts / total counts per cell) * scale.factor )

combined <- NormalizeData(combined, normalization.method = "LogNormalize", scale.factor = 10000)

# head(combined[["RNA"]]$data) # Normalized counts is in layer data
```

```{r}
# Feature selection
# Identification of highly variable features (hvg's)
# features that are highly expressed in some cells, and lowly expressed in others.

combined <-  FindVariableFeatures(combined, selection.method = "vst", nfeatures = 2000)
combined

```


```{r}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(combined), 10)
top10

```
```{r}
combined <- ScaleData(combined, vars.to.regress = "percent.mt")
```

```{r}
# Run PCA
combined <- RunPCA(combined)
ElbowPlot(combined)

```


```{r}
# Constructs a graph based on the PCs and clusters is done in PC space
# Cluster The Cells
combined <- FindNeighbors(combined, dims = 1:10)
combined <- FindClusters(combined, resolution = 0.25)
```
  
```{r}
combined <- RunUMAP(combined, dims = 1:10) 
DimPlot(combined, reduction = "umap", label = TRUE)

```

```{r}
# save the object
saveRDS(combined, file = "data/combined.rds")

```



```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones

combined <- JoinLayers(combined)
combined.markers <- FindAllMarkers(combined,
                                   only.pos = TRUE,
                                   min.pct = 0.25,
                                   logfc.threshold = 0.5)

```


```{r}
# Groups results by cluster.
# Keeps only genes with log2 fold change > 1, i.e., genes strongly overexpressed in that cluster compared to others.

combined.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

```

```{r}
# get top 3 upregulated genes for cluster 7

top3_cluster7 <- combined.markers %>%
  filter(cluster == 7) %>%
  arrange(desc(avg_log2FC)) %>%
  head(3)

top3_cluster7
```

```{r}
VlnPlot(combined, features = c("SELE", "CADM3-AS1", "MEOX1"))
```

```{r}
DotPlot(combined, features = c("SELE", "CADM3-AS1", "MEOX1"))
```


```{r}
FeaturePlot(combined, features = c("SELE", "CADM3-AS1", "MEOX1"))
```

```{r}
combined.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1.5) %>%
    slice_head(n = 3) %>%
    ungroup() -> top3

DoHeatmap(combined, features = top3$gene) + NoLegend()
```



```{r}
# Assigning cell type identity to clusters
# Search top3 genes for each cluster in PanglaoDB
top3

```

```{r}
# Cluster 0: CLDN5, TMEM100, FCN3 - Endothelial Cells. 
# Cluster1: CD69, CD3D, CD3E - T Cells.
# Cluster2: S100A2, IGFBP2, SFN - Basal Cells.
# Cluster3:AIF1, LST1, CD163 - Dendritic Cells.
# Cluster4: SCGB3A2,CYB5A,WFDC2 - Pulmonary Alveolar Type 2 Cells.
# Cluster5: PIGR - cholangiocytes or Epithelial Cells?, C16orf89 - Pulmonary Alveolar Type 2 Cells, CP - hepatocytes or cholangiocytes?
# Cluster6: DCN, COL1A2, LUM -Fibroblasts
# Cluster7: S100A12, GCA,S100A8 - Dendritic Cells
# Cluster8: GZMB,KLRD1,PRF1 - NK Cells
# Cluster9: MZB1, IGHA2, DERL3 - Plasma Cells
# Cluster10: COL6A3, SCN7A, MMP2 - Fibroblasts
# Cluster11: C1QA, C1QB, C1QC: Dendritic Cells
# Cluster12: EDNRB, IL1RL1, EMCN  - Endothelial Cells
# Cluster13: AGER, CLDN18, SCEL - Pulmonary Alveolar Type 2 Cells
# .....  
```


```{r}

new.cluster.ids <- c(
  "Endothelial Cells",      # cluster 0
  "T Cells",       # cluster 1
  "Basal Cells",     # cluster 2
  "Dendritic Cells",                # cluster 3
  "Pulmonary Alveolar Type 2 Cells",            # cluster 4
  "Unknown",        # cluster 5
  "Fibroblasts",        # cluster 6
  "Dendritic Cells",        # cluster 7
  "NK Cells",        # cluster 8
  "Plasma Cells",        # cluster 9
  "Fibroblasts",       # cluster 10
  "Dendritic Cells",       # cluster 11
  "Endothelial Cells",       # cluster 12
  "Pulmonary Alveolar Type 2 Cells",       # cluster 13
  "Unknown",       # cluster 14
  "Unknown",       # cluster 15
  "Unknown",       # cluster 16
  "Unknown",       # cluster 17
  "Unknown",       # cluster 18
  "Unknown"        # cluster 19
)

```


```{r}
names(new.cluster.ids) <- levels(combined)
combined <- RenameIdents(combined, new.cluster.ids)
```

```{r}
DimPlot(combined, reduction = "umap", pt.size = 0.5, label = TRUE) 
```

```{r}
# Idents(combined) # give cell barcode and its assigned cell type

# Save the manual cluster assignment to a new metadata column
combined$manual_assignment <- Idents(combined)

```

```{r}
combined[[]] # manual_assignment column is added to the metadata 
```

```{r}
DimPlot(combined, group.by = "manual_assignment", reduction = "umap", label = TRUE, pt.size = 0.5)
```

